cdc/generation.cc

```diff
@@ -119,176 +123,197 @@ const std::vector<token_range_description>& topology_description::entries() cons
-            entry.sharding_ignore_msb = get_sharding_ignore_msb(*endpoint, gossiper);
+/*
+ * Represents a wrap around vnode..
+ * This means a token range (|start|, |end|].
+ * A wrap around vnode always contains both 2^63-1 and -2^63.
+ * For example (100, 10].
+ */
+struct wrap_around_vnode final {
+    dht::token start;
+    dht::token end;
+    /*
+     * Finds the first token in this vnode that belongs to shard |shard_idx|.
+     *
+     * If there is no token that belongs to shard |shard_idx| in this vnode,
+     * |end| is returned.
+     *
+     * The first token means the one that appears first on the ring when going
+     * from |start| to |end|.
+     * 'first token' is not always the smallest.
+     * For example, if in vnode (100, 10] only tokens 110 and 1 belong to
+     * shard |shard_idx| then token 110 is the first because it appears first
+     * when going from 100 to 10 on the ring.
+     */
+    dht::token find_first_token_for_shard(
+            const dht::sharder& sharder, size_t shard_idx) const {
+        auto t = next_token(start);
+        if (sharder.shard_of(t) != shard_idx) {
+            t = sharder.token_for_next_shard(start, shard_idx);
         }
-
-        spots_to_fill += entry.streams.size();
-    }
-
-    auto schema = schema_builder("fake_ks", "fake_table")
-        .with_column("stream_id", bytes_type, column_kind::partition_key)
-        .build();
-
-    auto quota = std::chrono::seconds(spots_to_fill / 2000 + 1);
-    auto start_time = std::chrono::system_clock::now();
-
-    // For each pair (i, j), 0 <= i < streams.size(), 0 <= j < streams[i].size(),
-    // try to find a stream (stream[i][j]) such that the token of this stream will get mapped to this stream
-    // (refer to the comments above topology_description's definition to understand how it describes the mapping).
-    // We find the streams by randomly generating them and checking into which pairs they get mapped.
-    // NOTE: this algorithm is temporary and will be replaced after per-table-partitioner feature gets merged in.
-    repeat([&] {
-        for (int i = 0; i < 500; ++i) {
-            auto stream_id = make_random_stream_id();
-            auto token = dht::get_token(*schema, stream_id.to_partition_key(*schema));
-
-            // Find the token range into which our stream_id's token landed.
-            auto it = std::lower_bound(tokens.begin(), tokens.end(), token);
-            auto& entry = entries[it != tokens.end() ? std::distance(tokens.begin(), it) : 0];
-
-            auto shard_id = dht::shard_of(entry.streams.size(), entry.sharding_ignore_msb, token);
-            assert(shard_id < entry.streams.size());
-
-            if (!entry.streams[shard_id].is_set()) {
-                --spots_to_fill;
-                entry.streams[shard_id] = stream_id;
-            }
+        if (!t.is_maximum()) {
+            // This means we have found a token for shard |shard_idx| before 2^63-1
```

Comment:
So the comment does not say the whole truth.
it could be that `t == 2^63-1` here.

Reply:
I guess I wasn't precise. I meant <= 2^63-1. Will remove the -1.